### ISSUE TITLE
[R} CRAN is unhappy about calls to stderr in source code

### ISSUE THREAD

**Issue by jameslamb** [microsoft/LightGBM#1909]
This issue is a blocker for #629 (the path to CRAN distribution of `LightGBM`.

## Overview

CRAN does not permit compiled code in packages to terminate R or directly write to stdout/stderr. See this NOTE that arises:

```
* checking compiled code ... NOTE
Note: information on .o files for x64 is not available
File 'E:/sandbox/LightGBM/lightgbm.Rcheck/lightgbm/libs/x64/lib_lightgbm.dll':
  Found '_exit', possibly from '_exit' (C)
  Found 'abort', possibly from 'abort' (C), 'runtime' (Fortran)
  Found 'exit', possibly from 'exit' (C), 'stop' (Fortran)
  Found 'printf', possibly from 'printf' (C)

Compiled code should not call entry points which might terminate R nor
write to stdout/stderr instead of to the console, nor use Fortran I/O
nor system RNGs. The detected symbols are linked into the code but
might come from libraries and not actually be called.
It is good practice to register native routines and to disable symbol
search.

See 'Writing portable packages' in the 'Writing R Extensions' manual.
```

This is a tough issue to solve because the package uses code shared between R and Python.  Solving it for LightGBM probably involves doing something similar to what the XGBoost folks had to do ([see here](https://github.com/dmlc/xgboost/commit/1495a43cea05071e9451d008addffa598e4e7341)).

## How to run R CMD CHECK

To run R CMD CHECK on the package, clone the repo and install from source [per these directions](https://github.com/Microsoft/LightGBM/blob/master/R-package/README.md), then build the package and run R CMD CHECK.

On Mac OS, this looks like running the following from your local clone of `LightGBM`:

```
export CXX=/usr/local/bin/g++-8
export CC=/usr/local/bin/gcc-8

# build the package
Rscript build_r.R

# Check
LGB_VERSION=$(cat R-package/DESCRIPTION | grep ^Version | tr -d "Version: ")
R CMD CHECK lightgbm_${LGB_VERSION}.tar.gz --as-cran | tee check.log | cat
```

## How to close this issue

Contributions must pass our existing CI tests to be accepted. A contribution that fixes this issue will additionally result in R CMD CHECK not returning the note above on Mac OS.

**Comment by guolinke** [comment]
@jameslamb 
I remember LightGBM R's package is a hacking solution, which actually doesn't call any R's functions. 
We choose this due to the license issues.
Maybe we can hack for the R's print as well?

**Comment by jameslamb** [comment]
@guolinke Yes we probably will have to do something similar to what XGBoost did (see the link above).

Can you share the history with me on the specific license issues you had to hack around?

**Comment by jameslamb** [comment]
thanks @guolinke

**Comment by Laurae2** [comment]
@jameslamb related partially also to #1440 (Rpritnf control)

**Comment by guolinke** [comment]
closed due to duplicate to #1440
